### PR TITLE
Sequencer amnesia fix

### PR DESCRIPTION
**Pull request by dahliamalkhi**
This PR is another intermediate-step towards fixing the sequencer failover amnesia bug.

I removed the checkpointing/recovery of sequencer state, and left it as dead-code in comments with "TODO" and "FIXME" annotations. This is so if we need to make use of parts of it we can do it quickly. It should be removed once  we have confidence in this fix.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java [hunk 1]
```diff
@@ -242,7 +242,7 @@ public void syncObjectUnsafe(long timestamp) {
         // and it belongs to this thread use that
         if (optimisticallyOwnedByThreadUnsafe()) {
             // If there are no updates, ensure we are at the right snapshot
-            if (optimisticStream.pos() == Address.NEVER_READ) {
+            if (optimisticStream.pos() == Address.NON_ADDRESS) {
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java]: Can we move to NON_ADDRESS change to another PR, I don't agree with this because the constant is being used in a lot of places where it means precisely that: that the stream/object/etc has never been read.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **19e16d3**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #558 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/6lQDRZmYky8Uk8LKG1T62GExZvnTQr08)

**Comment by no2chem** [comment]
The latest push still uses NON_ADDRESS in places where it should be NEVER_READ (a stream which has never returned a read result should return NEVER_READ - NON_ADDRESS is not clear at all!). I don't want to put add a semantic argument in this PR, plus all the NON_ADDRESS additions definitely conflict with an upcoming PR which implements spliterator.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **88f85a0**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #558 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/BigTEjP65rBqaspTPujKiN1V5h94DY0T)

File: runtime/src/main/java/org/corfudb/runtime/view/Address.java [hunk 2]
```diff
@@ -8,11 +8,34 @@
...
+    /**
+     * @param addr
+     * @return true is addr is a legitimate address; false for all flag non-address constants
+     */
+    public static boolean isAddress(long addr) { return addr >= 0; }
+
+    /**
+     * @return a constant which can be used as the base for address iterations
+     */
+    public static long minAddress() { return 0L; }
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/Address.java]: Shouldn't this be just a static field like MAX?

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java [hunk 3]
```diff
@@ -160,14 +161,14 @@ void mergeTransaction() {
     public List<SMREntry> previous() {
         writePos--;
 
-        if (writePos <= Address.NEVER_READ) {
-            writePos = Address.NEVER_READ;
+        if (writePos <= Address.maxNonAddress()) {
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java]: should this check be a utility of the Address class ? Why do  we need to expose this ?

File: runtime/src/main/java/org/corfudb/runtime/view/stream/ReplexStreamView.java [hunk 4]
```diff
@@ -159,7 +159,7 @@ public ILogData previous() {
         doPendingSeek(getCurrentContext(), Address.MAX);
 
         // If never read or only one entry, return null
-        if (getCurrentContext().streamPointer == Address.NEVER_READ ||
+        if (getCurrentContext().streamPointer <= Address.maxNonAddress() ||
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/view/stream/ReplexStreamView.java]: I think we should pull these comparisons into the Address class. The operators need not be exposed.

File: test/src/test/java/org/corfudb/integration/ServerRestartTest.java [hunk 5]
```diff
@@ -132,6 +132,37 @@ public void testSequencerRecoveryFailoverServerAndClient() throws Exception {
     }
 
     /**
+     * Test recovery with a failed over server and a new client.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testDoubleFailoverRecovery() throws Exception {
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: Great test. I think zeeshan can add more variations to this. Including writing and reading more data.

Comment (zalokhan) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: Isn't this test exactly same as the one above this? testSequencerFailoverServerAndClient.
Secondly, contrary to the name there is only one failover taking place in the test.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **7e0e849**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #558 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/UvTdXCeG6YRfKuRGOiTDdMrFJ8Olvq6l)
